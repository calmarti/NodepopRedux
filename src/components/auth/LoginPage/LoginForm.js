import T from "prop-types";

import useForm from "../../../hooks/useForm";

function LoginForm({ onLogin, isLoading }) {
  const {
    formValue: credentials,
    setFormValue,
  } = useForm({
    email: "",
    password: "",
    remember: false,
  });
  const { email, password, remember } = credentials;

  //Nota: función 'handleChange' específica para LoginForm para poder hacer el test de React-Testing-Library

  const handleChange = ({ target: { name, value} }) => {
    setFormValue((currentCredentials) => ({
      ...currentCredentials,
      [name]: value,
    }));
  };

  const handleSubmit = (ev) => {
    ev.preventDefault();
    onLogin(credentials); 
  };


  const disabledButton =
    isLoading || !credentials.email || !credentials.password;

  return (
    <form onSubmit={handleSubmit}>
      <label>
        <span>Email</span>
        <input name="email" value={email} onChange={handleChange} />
      </label>

      <label>
        <span>Password</span>
        <input
          type="password"
          name="password"
          value={password}
          onChange={handleChange}
        />
      </label>

      <label>
        <span>Remember password</span>
        <input
          type="checkbox"
          name="remember"
          checked={remember}
          onChange={handleChange}
        />
      </label>

      <button
        disabled={
          disabledButton
        }
      >
        Login
      </button>
    </form>
  );
}

LoginForm.propTypes = {
  onLogin: T.func.isRequired,
};

export default LoginForm;
