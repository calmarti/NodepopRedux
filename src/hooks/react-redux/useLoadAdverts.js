import { useSelector } from "react-redux";
import { uiSelector, loadAdvertsSelector } from "../../store/selectors";

const useLoadAdverts = () => {
  const adverts = useSelector(loadAdvertsSelector);
  const { isLoading, error } = useSelector(uiSelector);
  return { adverts, isLoading, error };
};

export default useLoadAdverts;