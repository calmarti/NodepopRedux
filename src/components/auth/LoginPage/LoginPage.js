import React from "react";
import T from "prop-types";
import LoginForm from "./LoginForm";
import { connect } from "react-redux";
import { authLogin, uiResetError } from "../../../store/actions";
import { uiSelector } from "../../../store/selectors";


export function LoginPage({ onLogin, isLoading, error, resetError }) {
  return (
    <>
      <LoginForm  onLogin={onLogin} isLoading={isLoading} />
      {isLoading && <p>Loading...</p>}
      {error && (
        <div onClick={resetError} style={{ color: "red" }}>
          {error.message}
        </div>
      )}
    </>
  );
}

LoginPage.propTypes = {
  onLogin: T.func.isRequired,
};

const mapDispatchToProps = (dispatch, ownProps) => {
  return {
    onLogin: (credentials) => dispatch(authLogin(credentials, ownProps)),
    resetError: () => dispatch(uiResetError()),
  };
};

const mapStateToProps = (state) => {
  return uiSelector(state);
};

const ConnectedLoginPage = connect(
  mapStateToProps,
  mapDispatchToProps
)(LoginPage);

export default ConnectedLoginPage;
