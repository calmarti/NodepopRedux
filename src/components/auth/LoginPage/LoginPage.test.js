import { LoginPage } from "./LoginPage";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { fireEvent, render } from "@testing-library/react";

//Test del componente LoginPage con React-Testing-Library

describe("Test of redux-connected LoginPage with React-Testing-Library", () => {
  test("should execute handleLogin", () => {
    const store = {
      getState: () => ({
        auth: false,
        ui: { isLoading: false, error: null },
        adverts: { loaded: false, data: [] },
      }),
      subscribe: () => {},
      dispatch: jest.fn(),
    };
    const isLoading = false;
    const error = null;
    const onLogin = jest.fn();
    const renderObj = render(
      <Provider store={store}>
        <LoginPage
          isLoading={isLoading}
          error={error}
          onLogin={onLogin}
        />
      </Provider>
    );
    const { getByLabelText, getByRole } = renderObj;
    const usernameField = getByLabelText("Email");
    const passwordField = getByLabelText("Password");
    const rememberField = getByLabelText(/^Remember/);
    const submitButton = getByRole("button");

    expect(submitButton.disabled).toBe(true);

    const email = "admin@admin.com";
    const password = "1234";
    const remember = false;

    fireEvent.change(usernameField, { target: { value: email } });
    fireEvent.change(passwordField, { target: { value: password } });
    fireEvent.change(rememberField, { target: { value: false } });

    expect(submitButton.disabled).not.toBe(true);

    fireEvent.click(submitButton);
    expect(onLogin).toHaveBeenCalledWith({
      email,
      password,
      remember,
    });
  });
});

//Snapshot tests de LoginPage

describe("LoginPage snapshot tests", () => {
  test("snapshot test with prop 'isLoading' equal to false, prop 'error' equal to null", () => {
    const store = {
      getState: () => ({
        auth: false,
        ui: { isLoading: false, error: null },
        adverts: { loaded: false, data: [] },
      }),
      dispatch: () => {},
      subscribe: () => {},
    };

    const snapshot_1 = renderer
      .create(
        <Provider store={store}>
          <LoginPage onLogin={()=>{}} />
        </Provider>
      )
      .toJSON();
    expect(snapshot_1).toMatchSnapshot();
  });

  test("snapshot test with prop 'isLoading' equal to true, prop 'error' equal to null", () => {
    const store = {
      getState: () => ({
        auth: false,
        ui: { isLoading: true, error: null },
        adverts: { loaded: false, data: [] },
      }),
      dispatch: () => {},
      subscribe: () => {},
    };
    const snapshot_2 = renderer
      .create(
        <Provider store={store}>
          <LoginPage onLogin={()=>{}}  />
        </Provider>
      )
      .toJSON();
    // console.log(tree);
    expect(snapshot_2).toMatchSnapshot();
  });
});

test("snapshot test with prop 'isLoading' equal to false, prop 'error' different from null", () => {
  const store = {
    getState: () => ({
      auth: false,
      ui: { isLoading: false, error: { message: "Unauthorized" } },
      adverts: { loaded: false, data: [] },
    }),
    dispatch: () => {},
    subscribe: () => {},
  };
  const snapshot_3 = renderer
    .create(
      <Provider store={store}>
        <LoginPage onLogin={()=>{}} />
      </Provider>
    )
    .toJSON();
  // console.log(tree);
  expect(snapshot_3).toMatchSnapshot();
});
