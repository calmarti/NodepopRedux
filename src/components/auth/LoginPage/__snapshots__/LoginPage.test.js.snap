// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot tests snapshot test with prop 'isLoading' equal to false, prop 'error' equal to null 1`] = `
<form
  onSubmit={[Function]}
>
  <label>
    <span>
      Email
    </span>
    <input
      name="email"
      onChange={[Function]}
      value=""
    />
  </label>
  <label>
    <span>
      Password
    </span>
    <input
      name="password"
      onChange={[Function]}
      type="password"
      value=""
    />
  </label>
  <label>
    <span>
      Remember password
    </span>
    <input
      checked={false}
      name="remember"
      onChange={[Function]}
      type="checkbox"
    />
  </label>
  <button
    disabled={true}
  >
    Login
  </button>
</form>
`;

exports[`LoginPage snapshot tests snapshot test with prop 'isLoading' equal to true, prop 'error' equal to null 1`] = `
<form
  onSubmit={[Function]}
>
  <label>
    <span>
      Email
    </span>
    <input
      name="email"
      onChange={[Function]}
      value=""
    />
  </label>
  <label>
    <span>
      Password
    </span>
    <input
      name="password"
      onChange={[Function]}
      type="password"
      value=""
    />
  </label>
  <label>
    <span>
      Remember password
    </span>
    <input
      checked={false}
      name="remember"
      onChange={[Function]}
      type="checkbox"
    />
  </label>
  <button
    disabled={true}
  >
    Login
  </button>
</form>
`;

exports[`snapshot test with prop 'isLoading' equal to false, prop 'error' different from null 1`] = `
<form
  onSubmit={[Function]}
>
  <label>
    <span>
      Email
    </span>
    <input
      name="email"
      onChange={[Function]}
      value=""
    />
  </label>
  <label>
    <span>
      Password
    </span>
    <input
      name="password"
      onChange={[Function]}
      type="password"
      value=""
    />
  </label>
  <label>
    <span>
      Remember password
    </span>
    <input
      checked={false}
      name="remember"
      onChange={[Function]}
      type="checkbox"
    />
  </label>
  <button
    disabled={true}
  >
    Login
  </button>
</form>
`;
